test-framework/junit5/src/main/java/io/quarkus/test/junit/AppMakerHelper.java

```diff
@@ -86,65 +75,75 @@ private PrepareResult createAugmentor(final Class<?> requiredTestClass, String d
-                                if (!c.isAnnotationPresent((Class<? extends Annotation>) profile.getClassLoader()
-                                        .loadClass(Alternative.class.getName()))) {
-                                    throw new RuntimeException(
-                                            "Enabled alternative " + c + " is not annotated with @Alternative");
-                                }
-                            } catch (ClassNotFoundException e) {
-                                throw new RuntimeException(e);
-                            }
-                        })
-                        .map(Class::getName)
-                        .collect(Collectors.joining(",")));
-            }
-            if (profileInstance.disableApplicationLifecycleObservers()) {
-                additional.put("quarkus.arc.test.disable-application-lifecycle-observers", "true");
-            }
-            if (profileInstance.getConfigProfile() != null) {
-                additional.put(LaunchMode.TEST.getProfileKey(), profileInstance.getConfigProfile());
-            }
-
-            //we just use system properties for now
-            //it's a lot simpler
-            // TODO this is really ugly, set proper config on the app
-            // Sadly, I don't think #42715 helps, because it kicks in after this code
-            configCleanup = RestorableSystemProperties.setProperties(additional)::close;
-        }
-
-        if (curatedApplication
-                .getApplicationModel().getRuntimeDependencies().isEmpty()) {
-            throw new RuntimeException(
-                    "The tests were run against a directory that does not contain a Quarkus project. Please ensure that the test is configured to use the proper working directory.");
-        }
 
         // TODO should we do this here, or when we prepare the curated application?
         // Or is it needed at all?
         Index testClassesIndex = TestClassIndexer.indexTestClasses(testClassLocation);
         // we need to write the Index to make it reusable from other parts of the testing infrastructure that run in different ClassLoaders
         TestClassIndexer.writeIndex(testClassesIndex, testClassLocation, requiredTestClass);
 
+        return getPrepareResult(requiredTestClass, curatedApplication, profile, testClassLocation);
+    }
+
+    static QuarkusTestProfile getQuarkusTestProfile(Class<? extends QuarkusTestProfile> profile,
+            Collection<Runnable> shutdownTasks)
+            throws InstantiationException, IllegalAccessException, InvocationTargetException, NoSuchMethodException {
+        if (profile == null) {
+            return null;
+        }
+        final QuarkusTestProfile profileInstance = new ClassCoercingTestProfile(profile.getConstructor()
+                .newInstance());
+        final Map<String, String> additional = new HashMap<>();
+        // TODO we make this twice, also in abstractjvmextension can we streamline that?
```

Comment:
Does this comment still apply? I think you've just fixed that problem, and can delete the TODO.

Reply:
... or maybe the duplicated code has been fixed, but the execution still happens twice?

Reply:
The test profile is now created and checked in the `FacadeClassLoader.getOrCreateRuntimeClassLoader()` and the `QuarkusTestExtension.doJavaStart()`. I'll still need to look into how to optimize that.